google/cloud/pubsub/internal/streaming_subscription_batch_source.cc

```diff
@@ -0,0 +1,311 @@
+  });
+}
+
+void StreamingSubscriptionBatchSource::OnFinish(Status status) {
+  std::unique_lock<std::mutex> lk(mu_);
+  status_ = std::move(status);
+  stream_.reset();
+  ChangeState(lk, StreamState::kNull, __func__, "done");
+  if (shutdown_manager_->FinishedOperation("stream")) return;
+  lk.unlock();
+  shutdown_manager_->StartOperation(__func__, "stream",
+                                    [this] { StartStream(); });
+}
+
+void StreamingSubscriptionBatchSource::DrainQueues(
+    std::unique_lock<std::mutex> lk) {
+  if (ack_queue_.empty() && nack_queue_.empty() && deadlines_queue_.empty()) {
+    return;
+  }
+  if (stream_state_ != StreamState::kActive || pending_write_) return;
+  auto stream = stream_;
+  pending_write_ = true;
+
+  std::vector<std::string> acks;
+  acks.swap(ack_queue_);
+  std::vector<std::string> nacks;
+  nacks.swap(nack_queue_);
+  std::vector<std::pair<std::string, std::chrono::seconds>> deadlines;
+  deadlines.swap(deadlines_queue_);
+  lk.unlock();
+
+  google::pubsub::v1::StreamingPullRequest request;
+  for (auto& a : acks) request.add_ack_ids(std::move(a));
+  for (auto& n : nacks) {
+    request.add_modify_deadline_ack_ids(std::move(n));
+    request.add_modify_deadline_seconds(0);
+  }
+  for (auto& d : deadlines) {
+    request.add_modify_deadline_ack_ids(std::move(d.first));
+    request.add_modify_deadline_seconds(
+        static_cast<std::int32_t>(d.second.count()));
+  }
+  // Note that we do not use `AsyncRetryLoop()` here. The ack/nack pipeline is
+  // best-effort anyway, there is no guarantee that the server will act on any
+  // of these.
+  auto weak = WeakFromThis();
+  stream->Write(request, grpc::WriteOptions{}).then([weak](future<bool> f) {
+    if (auto self = weak.lock()) self->OnWrite(f.get());
+  });
+}
+
+void StreamingSubscriptionBatchSource::OnWrite(bool ok) {
+  std::unique_lock<std::mutex> lk(mu_);
+  pending_write_ = false;
+  if (ok && stream_state_ == StreamState::kActive) {
+    DrainQueues(std::move(lk));
+    return;
+  }
+  ShutdownStream(std::move(lk), ok ? "state" : "write error");
+}
+
+std::ostream& operator<<(std::ostream& os,
+                         StreamingSubscriptionBatchSource::StreamState s) {
+  using State = StreamingSubscriptionBatchSource::StreamState;
+  switch (s) {
+    case State::kNull:
+      os << "kNull";
+      break;
+    case State::kActive:
+      os << "kActive";
+      break;
+    case State::kDisconnecting:
+      os << "kDisconnecting";
+      break;
+    case State::kFinishing:
+      os << "kFinishing";
+      break;
+  }
+  return os;
+}
+
+void StreamingSubscriptionBatchSource::ChangeState(
+    std::unique_lock<std::mutex> const&, StreamState s, char const* where,
+    char const* reason) {
+  GCP_LOG(DEBUG) << where << " (" << reason << ") " << stream_state_ << ":" << s
+                 << " read=" << pending_read_ << " write=" << pending_write_
+                 << " shutdown=" << shutdown_ << " stream=" << stream_.get()
```

Comment:
Q. Is rendering an address meaningful?

Reply:
No. I just wanted to know if it was `nullptr`, changed the debug logging to do that. PTAL.